### ISSUE TITLE
Incorrect content when a process's published file is also an input and Fusion is enabled 

### ISSUE THREAD

**Issue by jordeu** [nextflow-io/nextflow#4309]
### Expected behavior and actual behavior

When using Fusion if a process published a file that is also an input of that process the content of the published file is just a Fusion path to the input file. This happens because this is the way of Fusion to manage symbolic links.

The expected behavior is that the published file is a copy of the input file, with same content. 

### Steps to reproduce the problem

```
$ nextflow run jordeu/nf-tests -r publish_input -w s3://fusion-develop/scratch
N E X T F L O W  ~  version 23.09.1-edge
Launching `https://github.com/jordeu/nf-tests` [stoic_rutherford] DSL2 - revision: b020907dd8 [publish_input]
executor >  local (fusion enabled) (2)
[b3/d224be] process > CREATE  [100%] 1 of 1 ✔
[b5/6c474f] process > PUBLISH [100%] 1 of 1 ✔

$ cat results/data.txt 
/fusion/s3/fusion-develop/scratch/b3/d224be3bcb4eefe89547058baba223/data.txt
```

### Environment 

```
$ nextflow info
  Version: 23.09.1-edge build 5881
  Created: 11-09-2023 10:08 UTC (12:08 CEST)
  System: Linux 5.15.0-83-generic
  Runtime: Groovy 3.0.19 on OpenJDK 64-Bit Server VM 17.0.4+8-LTS
  Encoding: UTF-8 (UTF-8)
```

**Comment by pditommaso** [comment]
Adding more context. The flow that causes the problem is the following: 

1. the process has an input named `foo.txt` with `stageAs` named "bar.txt"
2. the run script created by the nextflow creates a symlink eg `ln -s /fusion/s3/something/foo.txt bar.txt`  
3. Fusion creates a Fusion symbolic link, that means a file "bar.txt" with "/fusion/s3/something/foo.txt" as content.
4. the process has a publishDir of "bar.txt"
5. Nextflow is going to copy path "s3://workdir/.../bar.txt" into "s3://publishDir/.../bar.txt" without dereferencing it (as it is doing on grid platforms)
6. The user expects that the content of "s3://publishDir/.../bar.txt" is the same as "s3://something/foo.txt" but it is only a string (something like "/fusion/s3/something/foo.txt")

**Comment by pditommaso** [comment]
A possible solution could be to make Nextflow aware of the Fusion link, so that instead of copy that file, it follows the link. 

Those files are resolved in the snipper below. However this component is not aware that the file may be a fusion symlink, without other metainfo 

https://github.com/nextflow-io/nextflow/blob/89ede15c71e79858560f93df71064b74d2da2cf4/modules/nextflow/src/main/groovy/nextflow/processor/TaskProcessor.groovy#L1585-L1602

**Comment by pditommaso** [comment]
Also it must be said that by default input files are not included in the set of output files. So not sure why this is happening  

https://github.com/nextflow-io/nextflow/blob/89ede15c71e79858560f93df71064b74d2da2cf4/modules/nextflow/src/main/groovy/nextflow/processor/TaskProcessor.groovy#L1714-L1722

**Comment by jordeu** [comment]
> Also it must be said that by default input files are not included in the set of output files. So not sure why this is happening

It's because the input file is also define as output. See my example:

https://github.com/jordeu/nf-tests/blob/b020907dd8f8794090f884c9374cc0ce578c92f9/main.nf#L15-L28

**Comment by pditommaso** [comment]
So the `stageAs` is not the real problem

**Comment by pditommaso** [comment]
Adding example for reference 

```
process PUBLISH {
    publishDir "${params.outdir}"

    input:
      path "data.txt"

    output:
      path "data.txt"

    script:
    '''
    echo BYE >> data.txt
    '''
}
```

**Comment by pditommaso** [comment]
Worth mentioning @jordeu is also suggesting to have Fusion annotating object storage virtual links with some metadata that could be used by nextflow to resolve the link. 

Can you please provide more details ?

**Comment by jordeu** [comment]
I mean that Fusion could store the target path of the symbolic link as an object user metadata at `com.amazonaws.services.s3.model.ObjectMetadata`. 

The main problem that I see is that the metadata is not part of the [S3ObjectSummary](https://github.com/nextflow-io/nextflow/blob/b75ec444b02e18c57935de4bf79db4cc3f0a5644/plugins/nf-amazon/src/main/nextflow/cloud/aws/nio/S3Path.java#L523-L527) and it will require an extra request to S3 for each path that we resolve at publish dir [here](https://github.com/nextflow-io/nextflow/blob/62686ce86a8d8f24706fbd57b435e91ab4288891/modules/nextflow/src/main/groovy/nextflow/processor/PublishDir.groovy#L332-L348).

To reduce the performance impact of this lookup we could check for the existence of `.fusion.symlinks` (it contains a list of which files are symbolic links) and then only do this extra lookup for the paths that we know are symbolic links.

**Comment by pditommaso** [comment]
I believe it could be managed entirely on nextflow side without extra metadata

**Comment by jordeu** [comment]
Yes, instead of getting the target path from the metadata you can just read the contents of the file and extract the target path from that. Also this would be a more general solution that will work in the same way on all the object storage

**Comment by pditommaso** [comment]
is there a special syntax/format to recognise a link file?

**Comment by jordeu** [comment]
On the process folder there is a `.fusion.symlinks` file that contains the name (one per line) of all the files that are symbolic links, and then the content of the file is the target path.

Example:
```
$ aws s3 ls s3://fusion-develop/scratch/e7/8a6436c000c8ac180617a28596cc31/
2023-08-04 13:01:13          0 
2023-08-04 13:01:28          0 .command.begin
2023-08-04 13:01:28          0 .command.err
2023-08-04 13:01:30          5 .command.log
2023-08-04 13:01:30          5 .command.out
2023-08-04 13:01:13       3653 .command.run
2023-08-04 13:01:13         46 .command.sh
2023-08-04 13:01:30          1 .exitcode
2023-08-04 13:01:30      45245 .fusion.log
2023-08-04 13:01:30          6 .fusion.symlinks
2023-08-04 13:01:30         74 output

$ aws s3 cp s3://fusion-develop/scratch/e7/8a6436c000c8ac180617a28596cc31/.fusion.symlinks - 
output

$ aws s3 cp s3://fusion-develop/scratch/e7/8a6436c000c8ac180617a28596cc31/output -
/fusion/s3/fusion-develop/scratch/60/1aabf0626b92c2b75b4202a1ac0ad6/output

```

**Comment by pditommaso** [comment]
Understand, I was saying the content of the file that's supposed to be a link

**Comment by jordeu** [comment]
It's just a single line with the target Fusion path. In the example just this `/fusion/s3/fusion-develop/scratch/60/1aabf0626b92c2b75b4202a1ac0ad6/output`.

**Comment by pditommaso** [comment]
Maybe it could use special prefix to allow nextflow to easily recognise it without reading all file content. 

However, I still believe links can be inferred entirely using Nextflow metadata

**Comment by jordeu** [comment]
> However, I still believe links can be inferred entirely using Nextflow metadata

I never though on this approach. You mean to only fix the problem if the "symbolic links" are the inputs, the ones created by Nextflow, isn't it?

**Comment by bentsherman** [comment]
Draft PR here: #4348

It resolves the Fusion symlink correctly, but the output file is incorrect:
```console
$ cat results/data.txt
HELLO
```

Instead it should be:
```
HELLO
BYE
```

Both the output file source and the published file are incomplete.

**Comment by pditommaso** [comment]
@bentsherman can you please a unit test to validate it? [this](https://github.com/nextflow-io/nextflow/issues/4309#issuecomment-1721289033) should be enough

**Comment by bentsherman** [comment]
I have tested it with an e2e example and it works

**Comment by pditommaso** [comment]
I know, but better to keep under control over future changes/refactoring

**Comment by bentsherman** [comment]
oh, you mean add an integration test to the PR?

There is one complication:

> It resolves the Fusion symlink correctly, but the output file is incorrect:
> 
> ```
> $ cat results/data.txt
> HELLO
> ```
> 
> Instead it should be:
> 
> ```
> HELLO
> BYE
> ```
> 
> Both the output file source and the published file are incomplete.

If I add an integration test, I will have to make the test incorrect for it to pass. Then when the problem is fixed in Fusion, it will immediately break the Nextflow CI 😆

**Comment by bentsherman** [comment]
I have an idea... instead of having each process append a line to the same file, each process could create a new file and emit the accumulated set of files. That is a more valid use case anyway.

Meanwhile, it should probably be noted somewhere that modifying process inputs is an anti-pattern that does not currently work with Fusion.

**Comment by jordeu** [comment]
> I have an idea... instead of having each process append a line to the same file, each process could create a new file and emit the accumulated set of files. That is a more valid use case anyway.

You can also make the second process just read the symbolic linked input and forward it as output. In fact this is the pattern that I've seen on some reported tickets. 

> Meanwhile, it should probably be noted somewhere that modifying process inputs is an anti-pattern that does not currently work with Fusion.

This is a known issue scheduled to be fixed for the upcoming data studio support

**Comment by pditommaso** [comment]
Essentially just this 

```
process PUBLISH {
    publishDir "${params.outdir}"

    input:
      path "data.txt"

    output:
      path "data.txt"

    script:
    '''
    echo BYE 
    '''
}


```
